libcasr/src/rust.rs

```diff
@@ -26,3 +29,300 @@ impl Exception for RustPanic {
         )))
     }
 }
+
+/// Structure provides an interface for processing stacktraces from RUST_BACKTRACE.
+pub struct RustStacktrace;
+
+impl ParseStacktrace for RustStacktrace {
+    fn extract_stacktrace(stream: &str) -> Result<Vec<String>> {
+        let lines = stream
+            .split('\n')
+            .map(|l| l.to_string())
+            .collect::<Vec<String>>();
+
+        let Some(backtrace_idx) = lines
+            .iter()
+            .enumerate()
+            .position(|(i, line)| line.starts_with("stack backtrace:") && i < lines.len() - 1)
+        else {
+            return Err(Error::Casr(
+                "Couldn't find start of stacktrace in Rust panic output".to_string(),
+            ));
+        };
+
+        let lines = &lines[backtrace_idx + 1..];
+        let re_func = Regex::new(r" *\d:.+0x[0-9a-f]+ - .*").unwrap();
+        let re_src = Regex::new(r" *at .+").unwrap();
+        let mut stacktrace = Vec::new();
+        let mut st_line = String::new();
+        for line in lines {
+            if re_func.is_match(line) {
+                // New bt entry
+                if !st_line.is_empty() {
+                    // Entry can be with out sources
+                    stacktrace.push(st_line.clone());
+                    st_line.clear();
+                }
+                st_line.push_str(line.trim());
+            } else if re_src.is_match(line) {
+                // Extract sources
+                st_line.push(' ');
+                st_line.push_str(line.trim());
+                stacktrace.push(st_line.clone());
+                st_line.clear();
+            } else {
+                // Backtrace is finished
+                if !st_line.is_empty() {
+                    stacktrace.push(st_line.clone());
+                }
+                break;
+            }
+        }
+
+        if stacktrace.is_empty() {
+            return Err(Error::Casr(
+                "Couldn't find stacktrace entries in Rust panic output".to_string(),
+            ));
+        }
+
+        Ok(stacktrace)
+    }
+
+    fn parse_stacktrace_entry(entry: &str) -> Result<StacktraceEntry> {
+        let re =
+            Regex::new(r#" *[0-9]+: +0x([0-9a-f]+) - (.+) at (.+):([0-9]+):([0-9]+)"#).unwrap();
```

Comment:
you can cover two cases at once with regex `" *[0-9]+: +0x([0-9a-f]+) - (.+)\n( *at (.+):([0-9]+):([0-9]+))?"`